### DISCUSSION TITLE
How can I tell whether tree-sitter is actually performing an incremental parse?

### DISCUSSION THREAD

**Discussion by martaver** [tree-sitter/tree-sitter#1964] [Q&A]
I have a simple grammar that I'm using as a test bench to figure out incremental parsing:
```js
module.exports = grammar({
    name: 'my-grammar',
  
    rules: {                                

        source_file: $ => repeat(
            choice(
                $.service_declaration,
            )
        ),
        
        service_declaration: $=> seq('service', $.service_name),
    
        service_name: $ => /[_a-zA-Z\-][\w_]*/,
    }
  });
```

I'm using `web-tree-sitter` to get a parser, I load my language and try an incremental parse:
```js
const fromCode = `service foo service baz`;
const toCode = `service foo service bar service baz`;

// Parse the original text
const from = parser.parse(fromCode);

// Describe an edit that inserts `service bar` between `service foo` and `service baz`
const insertAt = 12;
const insertEnd = insertAt + toCode.length - fromCode.length;
from.edit({
    startIndex: insertAt,
    oldEndIndex: insertAt,
    newEndIndex: insertEnd,
    startPosition: { row: 0, column: insertAt },
    oldEndPosition: { row: 0, column: insertAt },
    newEndPosition: { row: 0, column: insertEnd }
})

// Try an incremental parse, 
const to = parser.parse(toCode, from);
```

Inspecting the `from` and `to` trees, though, I noticed the 'to' tree is correctly formed, but:
 - the 'from' tree's nodes that represented `service baz` now have empty text.
 - none of the node ids are re-used in 'to'... they all seems to have been recreated.

Also, changing the edit indices to ridiculous values doesn't produce a correspondingly messed up `to` tree... I guess if there are problems with the edit range tree-sitter tries to recover or just parse the full text.

Either way, it makes it kind of difficult to figure out what exactly is going on. I can't quite tell if tree-sitter is actually doing an incremental parse, either.

Can anyone help me understand how incremental parsing works?

**Comment by martaver** [comment]
Some progress! Regarding:

> the 'from' tree's nodes that represented service baz now have empty text.

...the old tree's nodes are empty, because their startIndex and endIndex have been updated, and are now out of bounds of the string that was supplied to the tree (which can be seen at rootNode.text).

No progress trying to meaningfully compare nodes from the old tree to the new tree, yet... occasionally some nodes share id but I don't see the logic in it. I would have expected all nodes related to text that's unchanged to have remained the same - but most of those unchanged nodes' ids change also. They are also never the same node when compared through referential equality.

**Reply by jmehnle** [reply]
FWIW, I observed the same issue with Tree Sitter's Python binding: once you call `tree.edit()` (not even `node.edit()`), all nodes you're holding a reference to immediately start returning an `None`-valued `text`.
